### PR TITLE
Add maven-failsafe-plugin and separate unit from integration tests

### PR DESCRIPTION
**Pull request by DanielFran**
See https://github.com/jhipster/generator-jhipster/issues/8868

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/server/templates/pom.xml.ejs [hunk 1]
```diff
@@ -1082,38 +1105,37 @@
...
-                                    <exclude>**/Abstract*.java</exclude>
-                                    <exclude>**/*Cucumber*.java</exclude>
-                                </excludes>
-                            </configuration>
-                        </execution>
-                        <execution>
-                            <id>without-cucumber</id>
                             <goals>
-                                <goal>test</goal>
+                                <goal>integration-test</goal>
```
[the remark this answers is not in the document]
Comment (DanielFran) [reply generators/server/templates/pom.xml.ejs]: Yes, will remove it!

File: generators/server/templates/pom.xml.ejs [hunk 2]
```diff
@@ -1082,38 +1105,37 @@
...
                         </execution>
                     </executions>
-                    <%_ } _%>
-                    <configuration>
-                        <!-- Force alphabetical order to have a reproducible build -->
-                        <runOrder>alphabetical</runOrder>
-                        <!-- Fixes a bug which will be solved in next version after 2.22.1 then this can be removed -->
-                        <!-- see: https://issues.apache.org/jira/browse/SUREFIRE-1588 -->
-                        <useSystemClassLoader>false</useSystemClassLoader>
-                        <outputDirectory>${sonar.junit.reportsPath}</outputDirectory>
```
[the remark this answers is not in the document]
Comment (DanielFran) [reply generators/server/templates/pom.xml.ejs]: I keep it, see lines 1111 and 1125

**Comment by DanielFran** [comment]
@pascalgrimaud I needed to change script 23-package to run integration test.
Can you confirm it is OK?

File: generators/server/templates/pom.xml.ejs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DanielFran) [reply generators/server/templates/pom.xml.ejs]: I needed to add them again otherwise failsafe did not run integration tests

**Comment by pascalgrimaud** [comment]
Ok @DanielFran, no worry you can change it if needed

**Comment by DanielFran** [comment]
@pascalgrimaud I believe my changes are OK, but actually in package we have errors!?!
I did not found why could be the root cause for the errors as these changes should not have impacts on package script...
Any clue?

**Comment by pascalgrimaud** [comment]
Thanks @DanielFran and @ruddell !
I think a fix would be to force the use of "terser": "3.14.1" for NPM and Yarn

**Comment by atomfrede** [comment]
I will take care of the required changes for gradle.

**Comment by pascalgrimaud** [comment]
I could have a look on it and test this PR, but should https://github.com/jhipster/generator-jhipster/pull/9173 be tested at the same time than your PR, @DanielFran ?

**Comment by pascalgrimaud** [comment]
Ok @DanielFran : self assigned, I'll test it asap

**Comment by pascalgrimaud** [comment]
@DanielFran : just tested right now, and it doesn't work.

- in my generator-jhipster project, `git checkout pr/9174` to use your branch
- I created an empty folder and go inside
- copy sample: `~/projects/jhipster/generator-jhipster/test-integration/generate-sample.sh generate $(pwd) ngx-default sql`
- generate the project with entities: `jhfe`
- start Sonar: `jhsonarup`
- backend tests: `./mvnw test`
- frontend tests: `npm test`
- launch Sonar analyze:
```
./mvnw sonar:sonar \
        -Dsonar.host.url=http://localhost:9001 \
        -Dsonar.projectKey=JHipsterSonar
```

Here the result:
capture d ecran de 2019-02-06 08-30-56

The coverage is very low, and Sonar doesn't display the unit tests.

**Comment by pascalgrimaud** [comment]
If I launch mvn verify, it looks fine.
I'll edit this comment later with additional informations.

**Comment by DanielFran** [comment]
@pascalgrimaud Since we separate unit from integration tests, launching mvn test will now just run unit tests (test classes that did not end with IntTest.java)... Will try to check it tomorrow night.
Note: Is it possible that you change sonar script to print the coverage?

**Comment by pascalgrimaud** [comment]
1) with this change, our sample project looks like having 0 unit test and it's bad in my opinion

2) do you mean the curl at the end of CI build ? it should work. At least, it works locally when I tested this PR

**Comment by pascalgrimaud** [comment]
@DanielFran : the PR https://github.com/jhipster/generator-jhipster/pull/9173 has been merged, so you need to resolve conflict.
Did you have a look on my previous comment ?

**Comment by DanielFran** [comment]
@pascalgrimaud Yes and I make the necessary changes in an project locally and it works, the issue was on https://github.com/jhipster/generator-jhipster/pull/9174/files#diff-46af807fe730aba11ead7c88e7e79a31R153, I needed to substitute <sonar.junit.reportPaths>${**project.testresult.directory**}/surefire-reports,${**project.testresult.directory**}/failsafe-reports</sonar.junit.reportPaths> by <sonar.junit.reportPaths>${**project.build.directory**}/surefire-reports,${**project.build.directory**}/failsafe-reports</sonar.junit.reportPaths> so it appears the total of unit + integration tests.

By mistake last days I delete the branch so this PR is orfan. :(

I created a new branch in my repository: https://github.com/DanielFran/generator-jhipster/tree/integration_tests

But yesterday I was blocked by this error https://travis-ci.com/DanielFran/generator-jhipster/jobs/177438895#L2927 and did not figure out what was the root cause.

**Comment by pascalgrimaud** [comment]
@DanielFran :
- in your fork, edit: `.git/config`
- edit only the `upstream` part with:
```
[remote "upstream"]
	url = https://github.com/jhipster/generator-jhipster
	fetch = +refs/heads/*:refs/remotes/upstream/*
	fetch = +refs/pull/*/head:refs/remotes/upstream/pr/*
```
- then, `git fetch upstream`
- get back your branch with: `git checkout pr/9174`

**Comment by pascalgrimaud** [comment]
About your error, I could have a look tomorrow if you're still blocked
Just tell me :)

**Comment by pascalgrimaud** [comment]
@DanielFran : do you need help on this ?

**Comment by DanielFran** [comment]
@pascalgrimaud Sorry I did not have time last weekend... will try today or tomorrow at last.

**Comment by DanielFran** [comment]
@pascalgrimaud I recreated from scratch the PR here https://github.com/DanielFran/generator-jhipster/commits/master but it seems not working in travis.

Using my local maven instalation it works.

**Comment by pascalgrimaud** [comment]
Ok, I'll have a look (so you push to your master branch)

**Comment by pascalgrimaud** [comment]
@DanielFran : it seems OK with Travis -> https://travis-ci.com/DanielFran/generator-jhipster/builds/101874265

**Comment by DanielFran** [comment]
@pascalgrimaud In fact if you see the logs you will see that surefire launch unit tests but failsafe did not launch integration tests, this is why coverage is lower...

Using my one maven + sonar everything is OK.
I suspect this might be due to maven binary ???

**Comment by DanielFran** [comment]
Closed in favor of https://github.com/jhipster/generator-jhipster/pull/9337

**Pull request closed without merging**